api/server/httputils/form.go

```diff
@@ -109,3 +114,31 @@ func ArchiveFormValues(r *http.Request, vars map[string]string) (ArchiveOptions,
 	}
 	return ArchiveOptions{name, path}, nil
 }
+
+// ParsePlatform parses the OCI platform from a form
+func ParsePlatform(form url.Values) (*ocispec.Platform, error) {
+	var p ocispec.Platform
+
+	p.OS = form.Get("os")
+	p.Architecture = form.Get("architecture")
+	p.Variant = form.Get("variant")
+	p.OSVersion = form.Get("osversion")
+	p.OSFeatures = form["osfeatures"]
```

Comment:
Wondering if we should include these... They seem to be ignored by default containerd matcher though: https://github.com/containerd/platforms/blob/c1438e911ac7596426105350652fe267d0fb8a03/platforms.go#L154-L159

On the other hand, I can image an index having multiple images for different OSVersions/Features and user wanting to push only a selected one.

So IMO, we should either:
1) Drop these parameters from the API
2) Use a custom platform matcher that also checks these

WDYT?

Reply:
Would it be horrible to use the canonical JSON representation in a single field instead of multiple fields?

Reply:
IMO, yes, because we pass the parameters by encoding them in url query, not as a json-encoded body :see-no-evil:

That doesn't really solve the `OSVersion` and `OSFeatures` issue though :P

Reply:
I would not be against JSON in url-encoding though if parsing that makes things less ambiguous (we already do this for filters, which admitted are a bit ugly, but in the end, API is not for humans)

Reply:
I think we specifically wanted to *not* use that string format in the API.

Reply:
That was my first implementation, see: https://github.com/moby/moby/pull/47679#discussion_r1585612288

Reply:
Changed to JSON encoding